### ISSUE TITLE
Confusing `enableSemaphoreDefaultBulkhead` property

### ISSUE THREAD

**Issue by OathMeadow** [spring-cloud/spring-cloud-circuitbreaker#224]
I am not sure if this is a bug, but It feels like one.

**Background**
As we know, by default, if no Bulkhead is configured, then `Spring Cloud Circuit Breaker` will use the `ThreadPoolBulkhead` for each circuit breaker. For us, we prefer semaphores for our event driven services, we have no need for thread pools. So, we found the property `spring.cloud.circuitbreaker.resilience4j.enableSemaphoreDefaultBulkhead` and we set it to `true`.

**Confusion**
For a service where all HTTP clients use a semaphore bulkhead, except one. What happens if the above property is set to `true` and `ThreadPoolBulkhead` is configured for the HTTP client `X`? The answer is a semaphore bulkhead!

This is the confusion, a `ThreadPoolBulkhead` configuration will never bite when the `enableSemaphoreDefaultBulkhead` is set to `true`. This property should instead be named `forceSemaphoreBulkhead`.

The code seems incorrect in my opinion. This parameter should `default` to a semaphore bulkhead, if no configuration says otherwise. The code I am talking about resides in the class `Resilience4jBulkheadProvider`.

```
	private boolean useSemaphoreBulkhead(String id) {
		return semaphoreDefaultBulkhead
				|| (bulkheadRegistry.find(id).isPresent() && threadPoolBulkheadRegistry.find(id).isEmpty());
	}
```

So basically the `semaphoreDefaultBulkhead`, in my opinion, should be checked if no `SemaphoreBulkhead` or `ThreadPoolBulkhead` exists, then default to a `SemaphoreBulkhead`, else a `ThreadPoolBulkhead`.

**Why even Mix?**
Well, the `ThreadPoolBulkhead` is preferable for our non-event driven services. Semaphores cannot be queued, only timed. But this applies only for certain clients. As it is now, we need to turn this flag off, and we have to explicitly configure all clients. 

Have we misunderstood this flag?

Thank you

**Comment by OathMeadow** [comment]
Something like this I had in mind:

```
private boolean useSemaphoreBulkhead(String id) {
    return bulkheadRegistry.find(id).isEmpty() && threadPoolBulkheadRegistry.find(id).isEmpty() 
            ? semaphoreDefaultBulkhead
            : threadPoolBulkheadRegistry.find(id).isEmpty();
}
```

However there is something I do not understand as it is now. If the flag `semaphoreDefaultBulkhead` is set to `false`, then we will never get any semaphore bulkhead, unless we manually register one in the `bulkheadRegistry`. In other words the `bulkheadRegistry.find(...)` will always return `false`, unless we manually register one via `resilience4jBulkheadProvider.addBulkheadCustomizer(...)`. 
Should the logic in `useSemaphoreBulkhead(String id)` only look for registered bulkheads or should it also look for registered configurations in the `bulkheadRegistry` / `threadPoolBulkheadRegistry`? It is possible that a certain circuit breaker is configured with a semaphore bulkhead in `application.yaml`. As it is now, the configuration will never matter because it will always default to a thread pool bulkhead.

Regarding configuration, I noticed that no custom configuration, except the defaults, are taken from the `bulkheadRegistry` or the  `threadPoolBulkheadRegistry`. It seems the only way to configure either is by doing a manual call to `Resilience4jBulkheadProvider.configure(...)`. It seems the `Resilience4jBulkheadProvider` only use a local configuration HashMap. In other words, no configuration done in the `application.yaml` will bite for either bulkhead type, except for `default`. Is this a bug, or am I misunderstanding something?

**Comment by ryanjbaxter** [comment]
I want to take a step back because this seems like there might be several different issues here and I am having a trouble keeping track of them.

This first one I see from your original post is this...

> For a service where all HTTP clients use a semaphore bulkhead, except one. What happens if the above property is set to true and ThreadPoolBulkhead is configured for the HTTP client X? The answer is a semaphore bulkhead!

If I understand this correctly you have `spring.cloud.circuitbreaker.resilience4j.enableSemaphoreDefaultBulkhead=true` but you have a `ThreadPoolBulkhead` configured for one specific service.  In this case we still use a semaphore bulkhead because `enableSemaphoreDefaultBulkhead=true`.  The problem is the logic in `useSemaphoreBulkhead`, currently it will always return `true` if `enableSemaphoreDefaultBulkhead=true` without checking first if the service have a threadpool bulkhead configured.  That I think I understand.

The second issue...

> Should the logic in useSemaphoreBulkhead(String id) only look for registered bulkheads or should it also look for registered configurations in the bulkheadRegistry / threadPoolBulkheadRegistry? It is possible that a certain circuit breaker is configured with a semaphore bulkhead in application.yaml. As it is now, the configuration will never matter because it will always default to a thread pool bulkhead.

The current logic of `useSemaphoreBulkhead` would return true if the circuit breaker is registered in the bulkhead registry and not in the threadpool registry, so if a circuit breaker was configured with a semaphore bulkhead in `application.yaml` it should use a semaphore bulkhead.

For the third isues...

>  It seems the Resilience4jBulkheadProvider only use a local configuration HashMap. In other words, no configuration done in the application.yaml will bite for either bulkhead type, except for default. Is this a bug, or am I misunderstanding something?

That isn't what I am seeing.  For example if I do this
`resilience4j.bulkhead.instances.slow.maxConcurrentCalls = 10` then I see a semaphore bulkhead used 
![CleanShot 2024-12-04 at 19 40 25@2x](https://github.com/user-attachments/assets/32bab8b4-6468-45e0-a59a-a28cd5f14647) and the configured bulkhead returned from the bulkhead registry in `Resilience4JBulkheadProvider.decorateCallable`

**Comment by OathMeadow** [comment]
Sorry If am confusing things. I will try to explain.

### The second issue...

If I were to configure the following:

```
resilience4j:
  bulkhead:
    configs:
      myCircuitBreaker:
        max-concurrent-calls: 5
```
The `myCircuitBreaker` is not registered inside the `Bulkhead` registry, but a config exist.
My application will, when needed, create the `myCircuitBreaker` via the `Resilience4JCircuitBreakerFactory::create(...)`. However, because no `Bulkhead` was registered, it will use the `ThreadPoolBulkhead` instead.

I take it you mean I have to instead configure:

```
resilience4j:
  bulkhead:
    instances:
      myCircuitBreaker:
        max-concurrent-calls: 5
```
I have not tested the above, but I assume it will be registered in the `BulkheadRegistry`. 
But the problem is, the other component like the `TimeLimiter` and `CircuitBreaker` works perfectly by just specifying a `config` rather than `instances` in the `application.yaml`. It is only `Bulkhead` we HAVE to pre-instantiate for `Resilience4jBulkheadProvider` to create a `Bulkhead` instead of a `ThreadPoolBulkhead`. This I think is confusing and inconsistent, and therefore perhaps of only checking registered `Bulkheads`, perhaps we should also consider any custom configuration.

I got fooled by this thinking I had a `Bulkhead` now once configured, just like the others, but the metrics told me I was still using a `ThreadPoolBulkhead`, that made me scratch my head.

### The third issue... (to be continued)

**Comment by ryanjbaxter** [comment]
When you use 

`resilience4j.bulkhead.configs` you are just specifying the configuration you would like to use for the bulkhead you are not actually creating a bulkhead.  To create the bulkhead you need to use `resilience4j.bulkhead.instances`.  

> But the problem is, the other component like the TimeLimiter and CircuitBreaker works perfectly by just specifying a config rather than instances in the application.yaml. 

Well a Bulkhead is still created without using `resilience4j.bulkhead.instances` it is just that the default bulkhead that is created is a threadpool bulkhead and not a semaphore bulkhead.  

>  It is only Bulkhead we HAVE to pre-instantiate for Resilience4jBulkheadProvider to create a Bulkhead instead of a ThreadPoolBulkhead.

Correct because by default we create threadpool bulkhead.  If you want to keep that as the default but want a specific bulkhead to use a semaphore bulkhead then it needs to be in the BulkheadRegistry and the only way to do that is to use `resilience4j.bulkhead.instances`.  Of course you can switch the default behavior by setting `enableSemaphoreDefaultBulkhead=true`.

Still looking into problem 3
